File: src/ts/backupEntryPolicy.test.ts
import { describe, expect, it } from "vitest";
import { classifyBackupEntry } from "@risuai/backup-core/entryPolicy.cjs";

describe("backup entry policy", () => {
  it("keeps core entries and supported asset layouts", () => {
    expect(classifyBackupEntry("database.risudat").kind).toBe("database");
    expect(classifyBackupEntry("encryption.risudat").kind).toBe("encryption");
    expect(classifyBackupEntry("avatar.png").kind).toBe("asset");
    expect(classifyBackupEntry("assets/avatar.png").kind).toBe("asset");
    expect(classifyBackupEntry("assets/folder/avatar.png").kind).toBe("asset");
  });

  it("keeps both cold-storage naming variants", () => {
    expect(classifyBackupEntry(
      "coldstorage/11111111-1111-1111-1111-111111111111.json",
    ).kind).toBe("coldStorage");
    expect(classifyBackupEntry(
      "coldstorage_22222222-2222-2222-2222-222222222222.json",
    ).kind).toBe("coldStorage");
  });

  it("treats unknown fork namespaces as optional extensions", () => {
    for (const name of [
      "inlay/abc.png",
      "inlay_sidecar/abc",
      "inlay_meta/abc",
      "other-fork/cache/item.bin",
    ]) {
      expect(classifyBackupEntry(name).kind).toBe("extension");
    }
  });

  it("rejects malformed or traversal paths", () => {
    expect(classifyBackupEntry("../escape.bin").kind).toBe("invalid");
    expect(classifyBackupEntry("fork//item.bin").kind).toBe("invalid");
    expect(classifyBackupEntry("/absolute.bin").kind).toBe("invalid");
  });
});

File: android/app/src/test/java/co/aiclient/risu/BackupEntryPolicyTest.java
package co.aiclient.risu;

import static org.junit.Assert.assertEquals;
import org.junit.Test;

public class BackupEntryPolicyTest {
    @Test
    public void classifiesCoreAndLegacyAssetEntries() {
        assertEquals(BackupEntryPolicy.Kind.DATABASE, BackupEntryPolicy.classify("database.risudat"));
        assertEquals(BackupEntryPolicy.Kind.ENCRYPTION, BackupEntryPolicy.classify("encryption.risudat"));
        assertEquals(BackupEntryPolicy.Kind.ASSET, BackupEntryPolicy.classify("avatar.png"));
        assertEquals(BackupEntryPolicy.Kind.ASSET, BackupEntryPolicy.classify("assets/avatar.png"));
        assertEquals(BackupEntryPolicy.Kind.ASSET, BackupEntryPolicy.classify("assets/folder/avatar.png"));
    }

    @Test
    public void recognizesColdStorageBeforeNamespaceFiltering() {
        assertEquals(BackupEntryPolicy.Kind.COLD_STORAGE, BackupEntryPolicy.classify(
            "coldstorage/11111111-1111-1111-1111-111111111111.json"
        ));
        assertEquals(BackupEntryPolicy.Kind.COLD_STORAGE, BackupEntryPolicy.classify(
            "coldstorage_22222222-2222-2222-2222-222222222222.json"
        ));
    }

    @Test
    public void ignoresUnknownForkNamespacesWithoutTreatingThemAsAssets() {
        assertEquals(BackupEntryPolicy.Kind.EXTENSION, BackupEntryPolicy.classify("inlay/abc.png"));
        assertEquals(BackupEntryPolicy.Kind.EXTENSION, BackupEntryPolicy.classify("inlay_sidecar/abc"));
        assertEquals(BackupEntryPolicy.Kind.EXTENSION, BackupEntryPolicy.classify("inlay_meta/abc"));
        assertEquals(BackupEntryPolicy.Kind.EXTENSION, BackupEntryPolicy.classify("another-fork/cache/item.bin"));
    }

    @Test
    public void rejectsTraversalAndMalformedPaths() {
        assertEquals(BackupEntryPolicy.Kind.INVALID, BackupEntryPolicy.classify("../escape.bin"));
        assertEquals(BackupEntryPolicy.Kind.INVALID, BackupEntryPolicy.classify("fork//item.bin"));
        assertEquals(BackupEntryPolicy.Kind.INVALID, BackupEntryPolicy.classify("/absolute.bin"));
    }
}
